### ISSUE TITLE
Data-generated test cases (parameterized tests)

### ISSUE THREAD

**Issue by bajtos** [mochajs/mocha#1454]
At the moment, generating multiple test cases that differ only by input data is rather cumbersome:

``` js
describe('prime number checker', function() {
  [2,3,5,7].forEach(function(value) {
    it('returns true for prime number ' + value, function() {
      expect(isPrime(value)).to.be.true();
    });

  [4,6,8,9].forEach(function(value) {
    it('returns false for composite number ' + value, function() {
      expect(isPrime(value)).to.be.false();
    });
});
```

I am proposing to extend the current API with syntactic sugar for data-generated test cases by adding an optional second parameter to both `it` and `describe`, this parameter will be an array of data points to test.

``` js
describe('prime number checker', function() {
  it('returns true for prime numbers', [2,3,5,7], function(data) {
    expect(isPrime(data)).to.be.true();
  });

  it('returns false for composite number', [4,6,8,9], function(data) {
    expect(isPrime(data)).to.be.false();
  });
});
```

More advanced example:

``` js
var SAMPLES = [
  { scheme: 'http', host: 'localhost', path: '/', url: 'http://localhost/' },
  { scheme: 'https', host: '127.0.0.1', path: '/test', url: 'https://127.0.0.1/test' }
];

describe('url helper', SAMPLES, function(data) {
  it('builds url from components', function() {
    var str = urlHelper.build({ scheme: data.scheme, host: data.host, path: data.path });
    expect(str).to.equal(data.url);
  });

  it('parses url into components', function() {
    var components = urlHelper.parse(data.url);
    expect(components).to.eql({ scheme: data.scheme, host: data.host, path: data.path });
  });
});
```

I am happy to contribute the implementation.

The purpose of this issue is to discuss necessary details and to get maintainers' approval before I start writing code.

Related to #57.

/cc @mcollina

**Comment by hellboy81** [comment]
TL;DR: how can I use parametrized tests?

**Comment by danielstjules** [comment]
Something like this can work :)

``` javascript
var assert = require('assert');

describe('suite', function() {
  [1, 2, 3].forEach(function(n) {
    it('correctly handles ' + n, function() {
      assert(n);
    });
  });
});

// =>

  suite
    ✓ correctly handles 1
    ✓ correctly handles 2
    ✓ correctly handles 3


  3 passing (7ms)
```

**Comment by hellboy81** [comment]
Very important: 

- `describe` can not be paramterized due on problems
   - reason?
- only `it`

**Comment by dasilvacontin** [comment]
> describe can not be paramterized due on problems

What problems?

```js
[1, 2, 3, 4].forEach(function (val) {
  describe('test ' + val, function () {
    it('bla', function () {

    })
  })
})
```

```
➜  js  mocha describe-parameterized.js


  test 1
    ✓ bla

  test 2
    ✓ bla

  test 3
    ✓ bla

  test 4
    ✓ bla


  4 passing (8ms)
```

**Comment by bananu7** [comment]
I agree with @dasilvacontin, this is the approach we use and it works reasonably well. Not sure if polluting the mocha's API with parametrization is necessary at all TBH.

**Comment by btelles** [comment]
FWIW, the @dasilvacontin 's solution only appears to work, but thte `val` value isn't changed in the `it` blocks:

    [1, 2, 3, 4].forEach(function (val) {
      describe('test ' + val, function () {
        it('bla', function () {
          console.log(val);
        })
      })
    })

    test 1
      ✓ bla
      4
    test 2
      ✓ bla
      4
    test 3
      ✓ bla
      4
    test 4
      ✓ bla
      4

**Comment by danielstjules** [comment]
@btelles What version of mocha are you using? Cause it works for me.

```javascript
$ cat test.js
[1, 2, 3, 4].forEach(function (val) {
  describe('test ' + val, function () {
    it('bla', function () {
      console.log(val);
    })
  })
})
dstjules:~/Desktop
$ mocha --version
2.4.5
dstjules:~/Desktop
$ mocha test.js


  test 1
1
    ✓ bla

  test 2
2
    ✓ bla

  test 3
3
    ✓ bla

  test 4
4
    ✓ bla


  4 passing (10ms)
```

**Comment by dasilvacontin** [comment]
@danielstjules Works here as well.

@btelles is that really the code you are using? From the output it looks as if you were referencing an iterator variable – once the `it`s are actually executed, the loop has finished and the iterator points/has the limit value.

**Comment by lawrencec** [comment]
FWIW, There's a module available I wrote called [Unroll](https://github.com/lawrencec/Unroll) which provide parameterized tests like the above but with named parameters which are also reflected in the test title. Useful if you want the test titles to be informative with regard to the parameters.

**Comment by mikejsdev** [comment]
I know this is old but there is now a really simple npm package to make this easier: [mocha-param](github.com/mikejsdev/mocha-param.git) 
![screen shot 2017-03-05 at 23 45 16](https://cloud.githubusercontent.com/assets/26206730/23592815/35b6ff80-01fe-11e7-819c-fcb5cbc4ee69.png)

**Comment by binduwavell** [comment]
@mikejsdev I don't understand the reasoning behind mocha-param when we can use: https://mochajs.org/#dynamically-generating-tests. The latter allows for unique names for each test, which is pretty huge. Presumably we can use this same dynamic technique to stamp out multiple describes too (although I have not tried that.) Can you help me understand the use cases where mocha-param is preferable to dynamically generating tests?

**Comment by timaschew** [comment]
I agree with @binduwavell 
And this has also been approved earlier by @dasilvacontin and @danielstjules and also by me now with mocha 3.2.0

At all people who writing modules and publishing them like @mikejsdev 
Please add a note about the motivation in your README:

- Why did you wrote the module?
- Which problems do you try to solve?
- What are you doing differently compared with X?

**Comment by OscarGalindo** [comment]
i think it's all about readability and maintainability, mocha-param for example is more readable than dynamically testing. 


` **tests.forEach(function(test) {**
    it(**'correctly adds ' + test.args.length + ' args'**, function() {
      var res = **add.apply(null, test.args);**
      assert.equal(res, test.expected);
    });
  });`

we should think that although this solves a problem, it can be improved

**Comment by fasttime** [comment]
Any way to do something like `it.only` and `it.skip` with (one particular instance of) a parameterized test? For me, that would be the only reason for using a helper package rather than dynamically generating tests.

**Comment by Overdrivr** [comment]
@binduwavell @timaschew 

Just my two cents, but coming from the Python world the syntax proposed in https://mochajs.org/#dynamically-generating-tests looks very poor to me:
1. The parameters must be retrieved through the `test` variable while in fact they should just be function parameters
2. Each `tests` entry must contain an object with key:value for each parameter. This makes the initial parametric data way more verbose than it should.
3. The test function itself (the one passed to `it(...)` has zero parameters which is not very explicit and readable

```
describe('add()', function() {
  var tests = [
   // See 2., extra key and values just for re-defining function parameters.
    {args: [1, 2], expected: 3},
    {args: [1, 2, 3], expected: 6},
    {args: [1, 2, 3, 4], expected: 10}
  ];

  tests.forEach(function(test) {
    it('correctly adds ' + test.args.length + ' args', 
      // See 3., no parameters to this function. Except it does have some
      function() {
      // See 1., cannot directly use `args` as a variable, need to go through `test.args`
      var res = add.apply(null, test.args);
      assert.equal(res, test.expected);
    });
  });
});
```

An (hypothetical) syntax that I would find superior would be:

```
describe('add()', function() {
  var tests = [
    // values, expected
    [[1, 2],  3],
    [[1, 2, 3],  6],
    [[1, 2, 3, 4], 10]
  ];

  using(tests).it('correctly adds with args', function(values, expected) {
      var res = add.apply(null, values);
      assert.equal(res, expected);
    });
  });
});
```
